### ISSUE TITLE
Common shared libraries no longer being packaged into deployment

### ISSUE THREAD

**Issue by n8sabes** [serverless/serverless#642]
v0.4.0 does not appear to be bundling Common Shared Libraries any longer.

```
project
|___meta
|__common_shared_libs
    |__db
    |__utils
|__component_a
    |__lib
    |__function_a
        |__handler.js <--- require("../../common_shared_libs/db")
|__component_b
    |__lib
    |__function_b
        |__handler.js <--- require("../../common_shared_libs/utils")
```

Upon deployment, the common_shared_libs does not seem to be coming along for the ride.

Workarounds such as:

1. SymLinks in the component directory can work, but can cause IDE confusion with multiple windows to the same code, or circular build rule issues.

2. Using `npm link target` or `npm --save link target` will not save into the package.json, so any new checkout of the project will require manual steps to run `npm link target` everywhere the common shared libraries are required.

3. Using `npm link target` method doesn't seem to work at all with Serverless Optimizer Plugin, which is critical for small production deployments.

4. Copying shared libs to each component is error prone due to code duplication.

It worked perfectly when the shared lib was under the /back directory. Am I overlooking something in my migration process?

**Comment by joostfarla** [comment]
Newer SLS versions zip up the whole component dir, but not any code outside of it. So you could consider changing your directory structure like this for example:

```
project
|___meta
|__component_x
    |__common_shared_libs
        |__db
        |__utils
    |__subcomponent_a
        |__lib
        |__function_a
            |__handler.js <--- require("../../common_shared_libs/db")
    |__subcomponent_b
        |__lib
        |__function_b
            |__handler.js <--- require("../../common_shared_libs/utils")
```

Furthermore, I would advice to use the [`serverless-optimizer-plugin`](https://github.com/serverless/serverless-optimizer-plugin) to bundle (including code outside of the component) and compress the code. This decreases the size of the Lambda code, which helps reducing cold startup times.

**Comment by n8sabes** [comment]
Thanks @joostfarla. I refactored the project as per your suggestion -- Two issues:

1) This hierarchy doesn't solve sharing a common library across "components". I really think we need to find a way to include shared libraries in a more effective manner (while still observing support for ES6 and other preprocessors / transpilers used throughout the project).

**More Importantly,**

2) The optimizer doesn't seem to work with this hierarchy as it tries to load node_module files from the shared-library based on my local dev system path. Here is the Lambda message reported back from an optimized shared-library function:

`errorMessage=Cannot find module '/Users/username/project/_meta/_tmp/function@1455492126549/shared-lib/node_modules/aws-sdk/apis/metadata.json'`

I posted this issue as a response on the optimizer project yesterday. It may be due to the node_modules in the shared-lib, not sure and need to investigate a bit more. I could try moving all the node modules into the parent component to see if it helps the optimizer -- but, that unfortunately completely removes the value of a shared-library.

Ideas welcome.
